stage0/src/sev.rs

```diff
@@ -104,18 +108,221 @@ pub fn deinit_ghcb(snp: bool, encrypted: u64) {
+    // Create a copy as we'll be consuming the range as we call `next()`, below.
+    let mut range = *range;
+
+    // A page table consists of 512 entries; let's figure out what the pages would be.
+    let start = Page::from_start_address(memory.page.start_address()).unwrap();
+    let pages = Page::<S>::range(start, start + 512);
+
+    // We have 512 entries to play with, but the frame range can contain many more entries than
+    // that. Thus, we need to iterate over the range in chunks of 512; the easiest way to do it is
+    // to iterate (mutably) over the page directory itself, and for each entry, fetch the next
+    // entry in the range. If we've covered everything, `range.next()` will return `None`, and
+    // the count will be zero once we've covered everything.
+    memory.pt.zero();
+    while memory
+        .pt
+        .iter_mut()
+        .filter_map(|entry| range.next().map(|frame| (entry, frame)))
+        .map(|(entry, frame)| {
+            entry.set_addr(
+                frame.start_address() + encrypted,
+                PageTableFlags::PRESENT | flags,
+            )
+        })
+        .count()
+        > 0
+    {
+        // We now know we've filled in at least one entry of the page table, and thus have something
+        // to validate.
+        tlb::flush_all();
+
+        if let Some((_, err)) = memory
+            .pt
+            .iter()
+            .zip(pages)
+            .filter(|(entry, _)| !entry.is_unused())
+            .map(|(entry, page)| (entry, page.pvalidate()))
+            .map(|(entry, result)| (entry, result.or_else(|err| f(entry.addr(), err))))
+            .find(|(_, result)| result.is_err())
+        {
+            return err;
+        }
+
+        // Clear out the page table, ready for the next iteration.
+        memory.pt.zero();
+    }
+
+    Ok(())
+}
+
+trait Validatable4KiB {
+    fn pvalidate(
+        &self,
+        pt: &mut MappedPage<Size2MiB>,
+        encrypted: u64,
+    ) -> Result<(), InstructionError>;
+}
+
+impl Validatable4KiB for PhysFrameRange<Size4KiB> {
+    fn pvalidate(
+        &self,
+        pt: &mut MappedPage<Size2MiB>,
+        encrypted: u64,
+    ) -> Result<(), InstructionError> {
+        pvalidate_range(
+            self,
+            pt,
+            encrypted,
+            PageTableFlags::empty(),
+            |addr, err| match err {
+                InstructionError::ValidationStatusNotUpdated => {
+                    let addr = addr.as_u64() & !encrypted;
+
+                    // Failing validation for addresses < 1 MiB is fine, as that memory was
+                    // validated in bootstrap assembly.
+                    if addr < 0x100000 {
+                        return Ok(());
+                    }
+
+                    log::warn!("validation status not updated for address {:#018x}", addr);
+                    Ok(())
+                }
+                other => Err(other),
+            },
+        )
+    }
+}
+
+trait Validatable2MiB {
```

Comment:
I am not sure, but it does seem a bit wrong to me to have two `traits` `Validatable2MiB` and `Validatble4Kb`. 

Why don't we have a field indicating the either `2MiB` or `4Kib` and match on this field?

Reply:
The main issue here is that the `pvalidate` signature is different, depending whether we have 2 MiB or 4 KiB pages. I don't particularly like it either, but refactoring this will take some thought.